File: elixir/02-variables/variables_patterns.exs
# Simple binding looks like assignment
x = 42
IO.puts("x = #{x}")

# Rebinding: the name can be bound to a new value
x = 100
IO.puts("x rebound to #{x}")

# Pattern matching with tuples
{a, b, c} = {1, "hello", :world}
IO.puts("a: #{a}, b: #{b}, c: #{c}")

# Pattern matching with lists
[head | tail] = [1, 2, 3, 4, 5]
IO.puts("head: #{head}")
IO.puts("tail: #{inspect(tail)}")

# Underscore ignores values you don't need
{_, second, _} = {:first, "I want this", :third}
IO.puts("second: #{second}")

# Pin operator (^) matches against existing value instead of rebinding
y = 10
{^y, z} = {10, 20}
IO.puts("y stayed: #{y}, z: #{z}")

# This would raise a MatchError because y is pinned to 10:
# {^y, z} = {99, 20}  # ** (MatchError) no match of right hand side value

# Pattern matching in function heads
defmodule Classifier do
  def describe(0), do: "zero"
  def describe(n) when n > 0, do: "positive"
  def describe(n) when n < 0, do: "negative"
end

IO.puts("\n--- Pattern Matching in Functions ---")
IO.puts("describe(0): #{Classifier.describe(0)}")
IO.puts("describe(42): #{Classifier.describe(42)}")
IO.puts("describe(-7): #{Classifier.describe(-7)}")

File: erlang/02-variables/variables_patterns.erl
#!/usr/bin/env escript
main(_) ->
    %% Basic binding
    X = 42,
    io:format("X = ~w~n", [X]),

    %% Matching against an already-bound variable succeeds if the value is the same
    X = 42,
    io:format("X still matches 42~n"),

    %% Rebinding X to a different value would crash:
    %% X = 100,  %% ** exception error: no match of right hand side value 100

    %% Pattern matching with tuples
    {A, B, C} = {1, hello, 3.14},
    io:format("A: ~w, B: ~w, C: ~w~n", [A, B, C]),

    %% Tagged tuples - common Erlang pattern
    {ok, Value} = {ok, "success"},
    io:format("Value: ~s~n", [Value]),

    %% Underscore ignores values
    {_, Second, _} = {first, "I want this", third},
    io:format("Second: ~s~n", [Second]),

    %% List pattern matching - head and tail
    [Head | Tail] = [1, 2, 3, 4, 5],
    io:format("Head: ~w~n", [Head]),
    io:format("Tail: ~w~n", [Tail]),

    %% Multiple elements from the head
    [First, Second2 | Rest] = [10, 20, 30, 40],
    io:format("First: ~w, Second2: ~w, Rest: ~w~n", [First, Second2, Rest]),

    %% Nested pattern matching
    {point, {PX, PY}} = {point, {100, 200}},
    io:format("Point X: ~w, Point Y: ~w~n", [PX, PY]),

    %% Using bound variables in patterns
    Target = error,
    {Target, Reason} = {error, "not found"},
    io:format("Matched ~w with reason: ~s~n", [Target, Reason]),

    %% Pattern matching in case expressions
    Result = {ok, 42},
    case Result of
        {ok, Val} ->
            io:format("~nCase matched ok: ~w~n", [Val]);
        {error, Err} ->
            io:format("~nCase matched error: ~w~n", [Err])
    end,

    ok.
